### PR TITLE
building an extended version of primekg that include OMIM phenotypes and genes

### PR DESCRIPTION
**Pull request by abearab**
Big thanks to @marinkaz for recommending a pull request to share my works here.

__Updates:__
- Including scripts to gather updated data from the OMIM portal using their API (credit to @amirieb)
- Including scripts to process and extract relevant information from downloaded OMIM datasets (downloaded on May 2023).
- Uploading a detailed notebook to build an extended version of PrimeKG (this can be a good example for users who want to append other resources to PrimeKG and build an extended version).
- A formatting update for the README file (I could add some notes about my work but I kept that for later).
 
__Also:__
- A discussion about these updates – https://github.com/mims-harvard/PrimeKG/issues/9
- https://github.com/mims-harvard/TDC/pull/207

### CODE REVIEW

**Comment by payalchandak** [comment]
Hi @abearab thank you for submitting this PR! I'm wondering if you are reading the OMIM information from a locally downloaded json file? It would be great if you could make this re-useable for others by including the source website and any wget commands! Thanks

**Comment by ayushnoori** [comment]
Thanks @abearab! We will review. Agree with @payalchandak, would be great to include the source information. Also, please feel free to add notes on your work to the README as part of this PR.

**Comment by abearab** [comment]
Hi @payalchandak and @ayushnoori

> I'm wondering if you are reading the OMIM information from a locally downloaded json file? It would be great if you could make this re-useable for others by including the source website and any wget commands!

I already added a notebook (see [this link](https://github.com/mims-harvard/PrimeKG/blob/2c11839e23aba458c3928ca7123a90f6f289223b/datasets/omim/omim-api.ipynb)) as part of this PR that describes how I made the local json file. Please let me know if that is what you meant or you need me to provide more info.

**Comment by abearab** [comment]
> Also, please feel free to add notes on your work to the README as part of this PR.
@ayushnoori plz see the updated README file. Let me know what you think. Thanks!

**Comment by abearab** [comment]
@payalchandak @ayushnoori

To give you an update, I just committed a few more features as a module in TDC. I believe it helps to build, handle, explore, and integrate knowledge graphs based on the final PrimeKG data format (i.e. simply a pandas data frame with same column names). This might be useful for future updates of PrimeKG itself and improves building graphs from other resources in a compatible format to PrimeKG.

https://github.com/mims-harvard/TDC/pull/207

___
Examples scripts:

```python
from tdc.knowledge_graph import KnowledgeGraph, build_KG

primekg = PrimeKG(path = './datasets/PrimeKG').to_KG()
```
1. List node from a given source
```python
primekg.get_nodes_by_source('NCBI').head()
```

id | type | name | source
-- | -- | -- | --
9796 | gene/protein | PHYHIP | NCBI
7918 | gene/protein | GPANK1 | NCBI
8233 | gene/protein | ZRSR2 | NCBI
4899 | gene/protein | NRF1 | NCBI
5297 | gene/protein | PI4KA | NCBI

2. Build a knowledge graph from scratch
```python
my_kg = build_KG(
    indices = REPLACE,# a list to assign row names of output data frame
    relation = REPLACE,# a list or string to assign values
    display_relation= REPLACE,# a list or string to assign values

    x_id = REPLACE,# a list or string to assign values
    x_type = REPLACE,# a list or string to assign values
    x_name = REPLACE,# a list or string to assign values
    x_source = REPLACE,# a list or string to assign values

    y_id = REPLACE,# a list or string to assign values
    y_type = REPLACE,# a list or string to assign values
    y_name = REPLACE,# a list or string to assign values
    y_source = REPLACE# a list or string to assign values
)
```

**Comment by ayushnoori** [comment]
@payalchandak, an update: @abearab and I were able to meet today. Abe let me know that there is no rush to merge this PR, since this PR is an extension of prior work done with Zak via SIBMI several years ago. Abe will keep developing this PR and also https://github.com/mims-harvard/TDC/pull/207, and will keep us posted.

@abearab, glad we were able to touch base today. Please feel free to keep myself and @payalchandak posted as your work progresses. As discussed, when we are ready to add OMIM phenotypes and genes to the KG, we will merge this PR. Thanks!

**Comment by payalchandak** [comment]
Thanks a lot @abearab and @ayushnoori! Looking forward to expanding PrimeKG further, please keep me posted.

**Comment by ayushnoori** [comment]
Hi @abearb, we're reviewing this PR. This looks great! A few comments and requests:

* May you please run `append_omim.ipynb` again but use `auxillary/kg_raw.csv`? This is the KG _before_ we take the largest connected component (please see [`build_graph.ipynb`](https://github.com/mims-harvard/PrimeKG/blob/main/knowledge_graph/build_graph.ipynb)), and it is best to add new edges to the KG before taking the LCC so we don't unnecessarily lose any nodes. This may resolve the `omim_genes_missing` problem that you note in `append_omim.ipynb`.

* It would be super helpful if you could prepare a summary of the changes to PrimeKG by adding OMIM nodes: _e.g._, the number of OMIM phenotypes already in PrimeKG, the number of OMIM phenotypes you add, the number of new edges added (stratified by edge type), the original vs. final edge counts, etc. Your Venn diagrams in `append_omim.ipynb` look fantastic – it would be helpful to have them in a single document with descriptions so we can review together before merging this PR.

Thanks so much! cc: @payalchandak @marinkaz

**Comment by abearab** [comment]
Hi @ayushnoori, Thanks for your reply here.

> * May you please run `append_omim.ipynb` again but use `auxillary/kg_raw.csv`? This is the KG _before_ we take the largest connected component (please see [`build_graph.ipynb`](https://github.com/mims-harvard/PrimeKG/blob/main/knowledge_graph/build_graph.ipynb)), and it is best to add new edges to the KG before taking the LCC so we don't unnecessarily lose any nodes. This may resolve the `omim_genes_missing` problem that you note in `append_omim.ipynb`.

Sounds good, I'll do this and get back to you.

> * It would be super helpful if you could prepare a summary of the changes to PrimeKG by adding OMIM nodes: _e.g._, the number of OMIM phenotypes already in PrimeKG, the number of OMIM phenotypes you add, the number of new edges added (stratified by edge type), the original vs. final edge counts, etc. Your Venn diagrams in `append_omim.ipynb` look fantastic – it would be helpful to have them in a single document with descriptions so we can review together before merging this PR.

Sure.

> * If data sharing policies permit, may you please also upload the OMIM source files used to construct the OMIM extension of the KG to Harvard Dataverse where we can link to them in the README? Please also feel free to create a new README file to direct users to the exact relative file path where these OMIM source files must be stored.

Please see this page – https://www.omim.org/api. I wanted to upload files into harvard dataverse on my end but I thought it may need approvals from OMIM so I stoped it to get some guidance here. Maybe @marinkaz already has that permission or she can tell us how we should proceed? I'll be happy to share the required files with you so you can upload it to the same "dataverse". Updating README file sounds good, I'll be happy to help with that, too.

I'll add more commits here and be happy to discuss more. Thanks

**Comment by abearab** [comment]
@ayushnoori plz see updates in [the README file](https://github.com/abearab/PrimeKG/blob/main/README.md) and [append_omim.ipynb](https://github.com/abearab/PrimeKG/blob/main/knowledge_graph/append_omim.ipynb) notebook. 

One note I would say here is that MONDO database I used here is most likely not same as what you are using. Thus, I'm skipping some node/edge annotations. Let me know if you have suggestion.

https://github.com/mims-harvard/PrimeKG/pull/12/commits/5489754ede803206be94d6362c6d706fcd31a609 I made subgraphs as mentioned earlier in a same format as current version of PrimeKG. Also, [this part](https://github.com/abearab/PrimeKG/tree/main?tab=readme-ov-file#dataloader-therapeutics-data-commons) can be updated once https://github.com/mims-harvard/TDC/pull/207 is merged.

___
cc @payalchandak

**Comment by ayushnoori** [comment]
Thanks for the updates, @abearab!

> May you please run append_omim.ipynb again but use auxillary/kg_raw.csv? This is the KG before we take the largest connected component (please see [build_graph.ipynb](https://github.com/mims-harvard/PrimeKG/blob/main/knowledge_graph/build_graph.ipynb)), and it is best to add new edges to the KG before taking the LCC so we don't unnecessarily lose any nodes. This may resolve the omim_genes_missing problem that you note in append_omim.ipynb.

Thinking about this – it would be great to make `append_omim.ipynb` a script which could be called by `build_graph.ipynb`. That is, when constructing the KG using `build_graph.ipynb`, the user should have the choice to add OMIM nodes _before_ taking the LCC, for example, by setting a variable flag `use_omim = True`.

Alternatively, you could modify `append_omim.ipynb` to start from `kg_raw.csv`, then copy the KG standardization steps from `build_graph.ipynb` into `append_omim.ipynb`.

> Thus, I'm skipping some node/edge annotations.

I'm not sure what you mean by this. Could you please provide a few examples? We should try and minimize node duplication between OMIM and MONDO as much as possible.

Also, in reviewing the README:
```
relation
mim_disease                        9599
mim_gene                          16636
mim_phenotype                    574128
mim_phenotypic_series              4111
mim_phenotypic_series_disease       549
phenotype_map                      7259
```

Can you please explain or provide examples for `phenotypic_series`, `phenotypic_series_disease`, and `phenotype_map` edge types? I look at the [OMIM documentation](https://www.omim.org/help/faq#1_13) but this wasn't clear to me.

> Please see this page – https://www.omim.org/api. I wanted to upload files into harvard dataverse on my end but I thought it may need approvals from OMIM so I stoped it to get some guidance here. 

I'll touch base with @marinkaz about sharing the OMIM-augmented KG publicly. In the interim, if you could privately share with us the OMIM-augmented KG files (_e.g._, via email), that would be great! We can then perform some standard checks on our end.

**Comment by ayushnoori** [comment]
Hi Abe, from Marinka:
> We can upload the files to https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/IXA7BM using a consistent naming scheme, such as `OMIM-augmented-edges.csv`, etc. The files should contain KG only, no derivative OMIM database, and absolutely no verbatim copies of any OMIM file. Abe should provide a notebook on how to parse the data and build a knowledge graph, but we cannot host the raw dataset in the repo or data verse.

If you send us (Payal, Marinka, and myself) the relevant files, we will upload them.

File: README.md
Comment (ayushnoori) [inline README.md:1]: Updated to fix typos and add clarity. LGTM!

File: datasets/omim/omim-api.ipynb
Comment (ayushnoori) [inline datasets/omim/omim-api.ipynb:1]: We might want to move this file to `datasets/processing_scripts/omim/omim-api.ipynb` for consistency.

File: datasets/processing_scripts/omim_tools.py
Comment (ayushnoori) [inline datasets/processing_scripts/omim_tools.py:1]: See comment above about moving to `/omim` for consistency.

File: datasets/omim/omim-api.ipynb
Comment (ayushnoori) [reply datasets/omim/omim-api.ipynb:1]: Also, can we make this into a Python script so we can add to `primary_data_resources.sh`?

File: knowledge_graph/append_omim.ipynb
Comment (ayushnoori) [inline knowledge_graph/append_omim.ipynb:1]: This looks really great! Thanks for the detailed documentation, @abearab.

**Review by ayushnoori — APPROVED** [review]
Few minor comments, but LGTM. Let's merge for now and we can work with @abearab to address and improve in later versions.

**Comment by ayushnoori** [comment]
@abearab, thanks for this great update to PrimeKG. Reviewed and merged.

Would be great if we could work together to address the minor points raised above (https://github.com/mims-harvard/PrimeKG/pull/12#issuecomment-1873095821) in a future update. For now, could you please send us the CSV files generated by your script (_e.g._, standalone OMIM edges and OMIM-augmented PrimeKG)? Thanks!

File: datasets/omim/omim-api.ipynb
Comment (abearab) [reply datasets/omim/omim-api.ipynb:1]: Surely it can be a Python script and be called in your `.sh` file. You would need to request an API Key from OMIM. I'm not sure how it can be relevant, but that can be added to this repository for automated actions https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions

File: datasets/processing_scripts/omim_tools.py
Comment (abearab) [reply datasets/processing_scripts/omim_tools.py:1]: I assumed this could be a place for keeping and adding more functions as "processing scripts" for OMIM. 

The notebook in `/datasets/omim/` is just been added for your record to show how I gathered the required datasets and data files will stay in that folder. But feel free to reorganize as you wish so 👍

**Comment by abearab** [comment]
Hi @ayushnoori 

Thanks for merging this!! plz see my responses below.

> > May you please run append_omim.ipynb again but use auxillary/kg_raw.csv? This is the KG before we take the largest connected component (please see [build_graph.ipynb](https://github.com/mims-harvard/PrimeKG/blob/main/knowledge_graph/build_graph.ipynb)), and it is best to add new edges to the KG before taking the LCC so we don't unnecessarily lose any nodes. This may resolve the omim_genes_missing problem that you note in append_omim.ipynb.
> 
> Thinking about this – it would be great to make `append_omim.ipynb` a script which could be called by `build_graph.ipynb`. That is, when constructing the KG using `build_graph.ipynb`, the user should have the choice to add OMIM nodes _before_ taking the LCC, for example, by setting a variable flag `use_omim = True`.
> 
> Alternatively, you could modify `append_omim.ipynb` to start from `kg_raw.csv`, then copy the KG standardization steps from `build_graph.ipynb` into `append_omim.ipynb`.
> 
> > Thus, I'm skipping some node/edge annotations.
> 
> I'm not sure what you mean by this. Could you please provide a few examples? We should try and minimize node duplication between OMIM and MONDO as much as possible.

To build links with MONDO I downloaded a file from MONDO website and I noticed some inconsistencies, also I didn't know how best to label nodes from MONDO. You may need to rerun `append_omim` notebook with your own local MONDO files for consistency.

Also, to my understanding, MONDO is not covering OMIM "gene" entries, and that was the main motivation for me to start this integration. I'm not sure what is the best way to avoid redundant MONDO and OMIM nodes in the final KG, I'm not very familiar with MONDO. 

> Also, in reviewing the README:
> 
> ```
> relation
> mim_disease                        9599
> mim_gene                          16636
> mim_phenotype                    574128
> mim_phenotypic_series              4111
> mim_phenotypic_series_disease       549
> phenotype_map                      7259
> ```
> 
> Can you please explain or provide examples for `phenotypic_series`, `phenotypic_series_disease`, and `phenotype_map` edge types? I look at the [OMIM documentation](https://www.omim.org/help/faq#1_13) but this wasn't clear to me.
> 
> > Please see this page – https://www.omim.org/api. I wanted to upload files into harvard dataverse on my end but I thought it may need approvals from OMIM so I stopped it to get some guidance here.

Phenotypic Series (PS) are OMIM efforts for charting phenotype sets, i.e. multiple OMIM phenotype pages will be grouped in one PS. For instance, this is something being used in UDN for patient phenotyping and diagnosis task purposes.

> I'll touch base with @marinkaz about sharing the OMIM-augmented KG publicly. In the interim, if you could privately share with us the OMIM-augmented KG files (_e.g._, via email), that would be great! We can then perform some standard checks on our end.

> Hi Abe, from Marinka:
> 
> > We can upload the files to https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/IXA7BM using a consistent naming scheme, such as `OMIM-augmented-edges.csv`, etc. The files should contain KG only, no derivative OMIM database, and absolutely no verbatim copies of any OMIM file. Abe should provide a notebook on how to parse the data and build a knowledge graph, but we cannot host the raw dataset in the repo or data verse.
> 
> If you send us (Payal, Marinka, and myself) the relevant files, we will upload them.

The `omim-api.ipynb` file in this PR is exactly what Marinka is asking for regarding "data gathering" and basic data cleaning related to our task here. If you get an OMIM API key you can run it yourself and get the most updated data from OMIM. The `append_omim.ipynb` is then how I extracted data from the resulting files downloaded from OMIM. 

I think this part is better to be revised with your local database, for instance, I have some assumptions about MONDO which can be naive or wrong. Here I'm skipping some information for `y` nodes that needs revision:
image

___
Anyway, I'll be happy to share any files if that's helpful. Just let me know if you need that from me.

Best

**Pull request merged**
